### PR TITLE
b/170497119: ordering route match#1 add uriTemplate

### PR DESCRIPTION
**Pull request by TAOXUY**
*Description*: import uriTemplate class from [path matcher filter](https://github.com/GoogleCloudPlatform/esp-v2/blob/641ce1d5c177401e424f2b27dd45de1bf797530b/src/api_proxy/path_matcher/http_template.h#L1). It is required to order route match sequence based on path matcher logic.

*Followup*:
- add `SequenceGenerator` under src/go/util/httppattern based on [path_matcher](https://github.com/GoogleCloudPlatform/esp-v2/blob/master/src/api_proxy/path_matcher/path_matcher.h)
- apply the route match ordering in `RouteGenerator`

### CODE REVIEW

**Review by qiwzhang — APPROVED** [review]
Looking very good.  Thanks

File: src/go/util/httppattern/uri_template.go [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+}
+
+const (
+	SingleParameterKey  = "/."
+	WildCardPathPartKey = "*"
+	WildCardPathKey     = "**"
+	InvalidChar         = byte(0)
+)
+
+// only constant path segments are allowed after '**'.
```
Comment (nareddyt) [inline src/go/util/httppattern/uri_template.go:100]: Are you sure about this, i think ** has to be last

https://cloud.google.com/endpoints/docs/grpc-service-config/reference/rpc/google.api#path-template-syntax

> The syntax ** matches zero or more URL path segments, which must be the last part of the URL path except the Verb.

File: src/go/util/httppattern/uri_template.go [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+func Parse(ht string) *UriTemplate {
+	if ht == "/" {
+		return &UriTemplate{}
+	}
+
+	p := parser{
+		input: ht,
+	}
+	if !p.parse() || !p.validateParts() {
+		return nil
```
Comment (nareddyt) [inline src/go/util/httppattern/uri_template.go:74]: We need some form of error reporting. At the very least, we can just return an error:

```
return nil, fmt.Errorf("Invalid http template (%v)", ht)
```

File: src/go/util/httppattern/uri_template.go [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+func (p *parser) consumeAllInput() bool {
+	return p.tb >= len(p.input)
+}
+
+func (p *parser) currentChar() byte {
+	if p.tb < p.te && p.te <= len(p.input) {
+		return p.input[p.te-1]
+	}
+
+	return InvalidChar
```
Comment (nareddyt) [inline src/go/util/httppattern/uri_template.go:343]: Callers of `currentChar()` do not handle this `InvalidChar`, is that expected? Will parsing succeed with these 0 characters in the template now?

File: src/go/util/httppattern/uri_template_test.go [hunk 4]
```diff
@@ -0,0 +1,1008 @@
...
+		"/a/b{x}c/s",
+		"/a/{x}b{y}/s",
+		// No verb test cases.
+		":",
+		"/:",
+		"/a/:",
+		"/a/*:",
+		"/a/**:",
+		"/a/{b=*}/**:",
+	}
```
Comment (nareddyt) [inline src/go/util/httppattern/uri_template_test.go:998]: A few more cases I'd like to verify:
- /a/**/b/*/c

File: src/go/util/httppattern/uri_template.go (hunk 3 above)
Comment (TAOXUY) [reply src/go/util/httppattern/uri_template.go:343]: Yes.

Nop, I think ""  are in the testcases.

File: src/go/util/httppattern/uri_template_test.go (hunk 4 above)
Comment (TAOXUY) [reply src/go/util/httppattern/uri_template_test.go:998]: Done.

File: src/go/util/httppattern/uri_template.go (hunk 2 above)
Comment (TAOXUY) [reply src/go/util/httppattern/uri_template.go:74]: The error info not under parsing environment will be meaningless and trivial. For example, 
- under such parsing stage, missing some components
- don't meet some expected part.

As for the `Invalid http template (%v)`, caller can add too if it is not successfully parsed.

File: src/go/util/httppattern/uri_template.go (hunk 1 above)
Comment (TAOXUY) [reply src/go/util/httppattern/uri_template.go:100]: According to the spec, it is invalid template.

I think it should be fine as an enhancement on template syntax. In addition, it is aligned to path matcher,  for which we don't want to break our users. WDYT? @qiwzhang

Comment (qiwzhang) [reply src/go/util/httppattern/uri_template.go:100]: I know our existing c++ http_template did not following that statement: ** has to be the last part.  Either way is fine for me.  If it is easy,  we just support it.  If it brings too much work for us,  we can drop it.

Comment (nareddyt) [reply src/go/util/httppattern/uri_template.go:100]: We should be careful about deviating from the spec. If we support something undocumented now, we have to keep supporting it forever.

If we do decide to support this, I would also make sure the grpc transcoding filter supports this, and update the docs.

**Review by nareddyt — APPROVED** [review]

File: src/go/util/httppattern/uri_template.go (hunk 1 above)
Comment (nareddyt) [reply src/go/util/httppattern/uri_template.go:100]: Well i guess if espv1 supports this (which is GA), we should support it to for making migration easy.

Comment (TAOXUY) [reply src/go/util/httppattern/uri_template.go:100]: ACK.

**Pull request merged**